classification/classification_network.py

```diff
@@ -246,109 +220,368 @@ def createClassifierRegion(network, classifierType, classifierParams):
+  input data.
+  @param regionName (str) name of the current region  
+  @param recordNumber (int) value of the current record number 
+  """
 
-  validateRegions(sensor, sp, tm, classifier)
+  region = network.regions[regionName]
+  region.setParameter("learningMode", True)
+  phaseInfo = ("-> Training '%s'. RecordNumber=%s. Learning is ON for %s, "
+               "but OFF for the remaining regions." % (regionName,
+                                                       recordNumber,
+                                                       trainedRegionNames))
+  print phaseInfo
 
-  linkRegions(network)
 
+
+def _stopLearning(network, trainedRegionNames, recordNumber):
+  """
+  Disable learning for all trained regions.
+  
+  @param network (Network) the network instance 
+  @param trainedRegionNames (list) regions that have been trained on the 
+  input data.
+  @param recordNumber (int) value of the current record number 
+  """
+
+  for regionName in trainedRegionNames:
+    region = network.regions[regionName]
+    region.setParameter("learningMode", False)
+
+  phaseInfo = ("-> Test phase. RecordNumber=%s. "
+               "Learning is OFF for all regions: %s" % (recordNumber,
+                                                        trainedRegionNames))
+  print phaseInfo
+
+
+
+def runNetwork(network, networkConfiguration, numRecords):
+  """
+  Run the network and write classification results output.
+  
+  @param networkConfiguration (dict) the configuration of this network.
+  @param numRecords (int) Number of records of the input dataset.
+  @param network (Network) a Network instance to run.
+  """
+
+  partitions = _findNumberOfPartitions(networkConfiguration, numRecords)
+  sensorRegion = network.regions[SENSOR_REGION_NAME]
+  classifierRegion = network.regions[CLASSIFIER_REGION_NAME]
+  testIndex = partitions[-1][1]
+
+  # keep track of the regions that have been trained
+  trainedRegionNames = []
+  numCorrect = 0
+  numTestRecords = 0
+  for recordNumber in xrange(numRecords):
+    # Run the network for a single iteration
+    network.run(1)
+
+    actualValue = sensorRegion.getOutputData("categoryOut")[0]
+
+    partitionName = partitions[0][0]
+    partitionIndex = partitions[0][1]
+
+    # train all of the regions
+    if partitionIndex == recordNumber:
+
+      if partitionName == TEST_PARTITION_NAME:
+        _stopLearning(network, trainedRegionNames, recordNumber)
+
+      else:
+        partitions.pop(0)
+        trainedRegionNames.append(partitionName)
+        _enableRegionLearning(network,
+                              trainedRegionNames,
+                              partitionName,
+                              recordNumber)
+
+    # Evaluate the predictions on the test set.
+    if recordNumber >= testIndex:
+      if classifierRegion.type == KNN_CLASSIFIER_TYPE:
+        inferredValue = classifierRegion.getOutputData("categoriesOut").argmax()
```

Comment:
`argmax()` is not a good way to get the inference value b/c it handles ties by always taking the last occurrence. Try this:

    # The use of numpy.lexsort() here is to first sort by labelFreq, then sort
    # by random values; this breaks ties in a random manner.
    inferenceValues = classifierRegion.getOutputData("categoriesOut")
    randomValues = numpy.random.random(inferenceValues.size)
    inferredValue = numpy.lexsort((randomValues, inferenceValues))[:-1]